Handle suppress-HDR notifications for videos

squelart opened the pull request:
#### 1285af2e9967b3e26c6c8b67a60f73d1b3f9594e
<pre>
Handle suppress-HDR notifications for videos
<a href="https://bugs.webkit.org/show_bug.cgi?id=288690">https://bugs.webkit.org/show_bug.cgi?id=288690</a>
<a href="https://rdar.apple.com/145717196">rdar://145717196</a>

Reviewed by Simon Fraser.

This patch observes notifications or trait changes that
request the application or view to suppress HDR content
in videos (by constraining it).

This signal gets combined with the existing CSS
`dynamic-range-limit` to enforce the minimum of the two
constraints.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _registerForNotifications]):
(-[WKWebView _UITraitHDRHeadroomUsageDidChange]):
Trait change entry point on iOS, forwarded to WebPageProxy.
The initial value is also sent.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setShouldSuppressHDR):
* Source/WebKit/UIProcess/WebPageProxy.h:
Forwarded to WebPage. (To be continued further below...)

* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):
(WebKit::WebProcessPool::unregisterNotificationObservers):
Notification entry point on macOS, forwarded to all WebProcess&apos;es.

* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::createWebPage):
(WebKit::WebProcess::setShouldSuppressHDR):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
Cache the shouldSuppressHDR flag (for future page creations),
and forward to WebPage&apos;s.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setShouldSuppressHDR):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Meeting point between iOS trait changes and macOS notifications.
Forward to Page.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::setShouldSuppressHDR):
* Source/WebCore/page/Page.h:
(WebCore::Page::shouldSuppressHDR const):
Store the flag (for use by document and elements).
Notify Document if the flag changed.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::shouldSuppressHDRDidChange):
* Source/WebCore/dom/Document.h:
Notify all media elements.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::computePlayerDynamicRangeLimit const):
(WebCore::HTMLMediaElement::dynamicRangeLimitDidChange):
(WebCore::HTMLMediaElement::shouldSuppressHDRDidChange):
* Source/WebCore/html/HTMLMediaElement.h:
On notification (or dynamic-range-limit change), combine Page&apos;s
suppress-HDR with dynamic-range-limit, and update MediaPlayer&apos;s
effective dynamic range limit (already implemented in previous patch).

* Source/WebKit/Platform/spi/mac/AppKitSPI.h:
Add include relevant to notifications.

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::effectiveDynamicRangeLimitValue):
(WebCore::Internals::setPageShouldSuppressHDR):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
Simulate notifications at the Page level, examine effective
dynamic range limit in media elements.

* LayoutTests/media/video-suppress-hdr-dynamic-range-limit-expected.txt: Added.
* LayoutTests/media/video-suppress-hdr-dynamic-range-limit.html: Added.
* LayoutTests/media/video-suppress-hdr-expected.txt: Added.
* LayoutTests/media/video-suppress-hdr.html: Added.
Test combinations of notifications and video creations.

Canonical link: <a href="https://commits.webkit.org/291383@main">https://commits.webkit.org/291383@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d185f00ae15fc3204f7645dc955ab12dcfe26307

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92780 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/12331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1972 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/43291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94830 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12611 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20783 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/97776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/43291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95782 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/83967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/1599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/42619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1570 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99798 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19832 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20083 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79861 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79321 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19670 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23849 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1119 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19816 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24992 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/19503 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22963 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/21244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -7906,6 +7906,13 @@ void HTMLMediaElement::markCaptionAndSubtitleTracksAsUnconfigured(ReconfigureMod
+{
+    if (!m_shouldSuppressHDR)
+        return m_platformDynamicRangeLimit;
+    return std::min(m_platformDynamicRangeLimit, PlatformDynamicRangeLimit::constrainedHigh());
```
Don't we want `noLimit` by default?

squelart replied:
In the `m_shouldSuppressHDR==true` case, we do want to limit to max `constrainedHigh`, even for videos.
Or did I get my logic wrong?

smfr replied:
This is OK.

smfr commented on Source/WebCore/html/HTMLMediaElement.h:
```diff
@@ -1342,6 +1344,7 @@ class HTMLMediaElement
 
     std::optional<DynamicRangeMode> m_overrideDynamicRangeMode;
     PlatformDynamicRangeLimit m_platformDynamicRangeLimit;
+    bool m_shouldSuppressHDR { false };
```
This PR doesn't appear to support the case when a video element is created while HDR is suppressed. We also need some way to test this.

squelart replied:
That case is now supported, as well as creating new pages.
I'm still working on adding tests...

squelart replied:
Now with tests.

heycam commented on Source/WebKit/WebProcess/WebProcess.messages.in:
```diff
@@ -75,6 +75,8 @@ messages -> WebProcess : AuxiliaryProcess WantsAsyncDispatchMessage {
 
     SetTextCheckerState(OptionSet<WebKit::TextCheckerState> textCheckerState)
 
+    setShouldSuppressHDR(bool shouldSuppressHDR)
```
This should start with a capital letter for consistency with other messages names.

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -2966,6 +2966,12 @@ void WebPage::setUnderPageBackgroundColorOverride(WebCore::Color&& underPageBack
+void WebPage::setShouldSuppressHDR(bool shouldSuppressHDR)
+{
+    RefPtr page = m_page;
+    page->setShouldSuppressHDR(shouldSuppressHDR);
```
`protectedCorePage()-> setShouldSuppressHDR`

squelart committed:
Handle suppress-HDR notifications for videos
https://bugs.webkit.org/show_bug.cgi?id=288690
rdar://145717196

Reviewed by Simon Fraser.

This patch observes notifications or trait changes that
request the application or view to suppress HDR content
in videos (by constraining it).

This signal gets combined with the existing CSS
`dynamic-range-limit` to enforce the minimum of the two
constraints.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _registerForNotifications]):
(-[WKWebView _UITraitHDRHeadroomUsageDidChange]):
Trait change entry point on iOS, forwarded to WebPageProxy.
The initial value is also sent.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setShouldSuppressHDR):
* Source/WebKit/UIProcess/WebPageProxy.h:
Forwarded to WebPage. (To be continued further below...)

* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::registerNotificationObservers):
(WebKit::WebProcessPool::unregisterNotificationObservers):
Notification entry point on macOS, forwarded to all WebProcess'es.

* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::createWebPage):
(WebKit::WebProcess::setShouldSuppressHDR):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
Cache the shouldSuppressHDR flag (for future page creations),
and forward to WebPage's.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setShouldSuppressHDR):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Meeting point between iOS trait changes and macOS notifications.
Forward to Page.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::setShouldSuppressHDR):
* Source/WebCore/page/Page.h:
(WebCore::Page::shouldSuppressHDR const):
Store the flag (for use by document and elements).
Notify Document if the flag changed.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::shouldSuppressHDRDidChange):
* Source/WebCore/dom/Document.h:
Notify all media elements.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::computePlayerDynamicRangeLimit const):
(WebCore::HTMLMediaElement::dynamicRangeLimitDidChange):
(WebCore::HTMLMediaElement::shouldSuppressHDRDidChange):
* Source/WebCore/html/HTMLMediaElement.h:
On notification (or dynamic-range-limit change), combine Page's
suppress-HDR with dynamic-range-limit, and update MediaPlayer's
effective dynamic range limit (already implemented in previous patch).

* Source/WebKit/Platform/spi/mac/AppKitSPI.h:
Add include relevant to notifications.

* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::effectiveDynamicRangeLimitValue):
(WebCore::Internals::setPageShouldSuppressHDR):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
Simulate notifications at the Page level, examine effective
dynamic range limit in media elements.

* LayoutTests/media/video-suppress-hdr-dynamic-range-limit-expected.txt: Added.
* LayoutTests/media/video-suppress-hdr-dynamic-range-limit.html: Added.
* LayoutTests/media/video-suppress-hdr-expected.txt: Added.
* LayoutTests/media/video-suppress-hdr.html: Added.
Test combinations of notifications and video creations.

Canonical link: https://commits.webkit.org/291383@main
